test/src/test/java/org/corfudb/runtime/clients/TestClientRouter.java

```diff
@@ -218,17 +276,99 @@ public IClientRouter addClient(IClient client) {
+        CorfuMessage.HeaderMsg header = CorfuProtocolMessage.getHeaderMsg(thisRequestId, priority,
+                epoch, clusterId, protoClientId, ignoreClusterId, ignoreEpoch);
+        CorfuMessage.RequestMsg request = CorfuMessage.RequestMsg.newBuilder().setHeader(header)
+                .setPayload(payload).build();
+
+        // Generate a future and put it in the completion table.
+        final CompletableFuture<T> cf = new CompletableFuture<>();
+        outstandingRequests.put(thisRequestId, cf);
+
+        // Evaluate rules.
+        if (rules.stream().allMatch(x -> x.evaluate(request, this))) {
+            log.trace(Thread.currentThread().getId() + ":Sent request: {}", request);
+            // Write the message out to the channel
+            try {
+                RequestMsg requestMsg = simulateSerialization(request);
+                serverRouter.sendServerMessage(requestMsg, channelContext);
+            } catch (IOException e) {
+                log.error("encode: Error during serialization or deserialization!", e);
+            }
+        }
+        // Generate a benchmarked future to measure the underlying request
+        final CompletableFuture<T> cfBenchmarked = cf.thenApply(x -> {
+            MetricsUtils.stopConditionalContext(roundTripMsgContext);
+            return x;
+        });
+
+        // Generate a timeout future, which will complete exceptionally if the main future is not completed.
+        final CompletableFuture<T> cfTimeout = CFUtils.within(cfBenchmarked, Duration.ofMillis(timeoutResponse));
+        cfTimeout.exceptionally(e -> {
+            outstandingRequests.remove(thisRequestId);
+            log.debug("Remove request {} due to timeout!", thisRequestId);
+            return null;
+        });
+        return cfTimeout;
+    }
+
     // Create a timer using appropriate cached timer names
     private Timer getTimer(@NonNull CorfuMsg message) {
         if (!timerNameCache.containsKey(message.getMsgType())) {
             timerNameCache.put(message.getMsgType(),
-                               CorfuComponent.CLIENT_ROUTER.toString() +
-                               message.getMsgType().name().toLowerCase());
+                    CorfuComponent.CLIENT_ROUTER.toString() +
+                            message.getMsgType().name().toLowerCase());
         }
         return CorfuRuntime.getDefaultMetrics()
                 .timer(timerNameCache.get(message.getMsgType()));
     }
 
+    // Create a timer using appropriate cached timer names
+    private Timer getProtoTimer(@NonNull RequestPayloadMsg.PayloadCase payloadCase) {
+        if (!protoTimerNameCache.containsKey(payloadCase)) {
+            protoTimerNameCache.put(payloadCase,
+                    CorfuComponent.CLIENT_ROUTER.toString() + payloadCase.name().toLowerCase());
+        }
+        return CorfuRuntime.getDefaultMetrics()
+                .timer(protoTimerNameCache.get(payloadCase));
+    }
+
```

Comment:
Why is a timer needed for the test router?

Reply:
Good point, I've cleaned it up.